### PR TITLE
String templating in API Services

### LINKED ISSUE #79: Add string templating and replace http url construction
- [x] Add code from the following: https://gist.github.com/atruskie/12abf8cc7979eec14fbee956f13e943c
- [x] replace the implentation of `refineUrl`  (workbench-client/src/app/services/baw-api/api-common.ts) with invocations of the template function

### PR DESCRIPTION
**Pull request by Allcharles**
- Replaced the implementation of `refineUrl` with `makeTemplate`
- Improved http request handling
  - Remove separate http type functions (get, post, patch, delete)
  - Created apiRequest function (list, filter, show, create, update, delete)

Closes: #79

### CODE REVIEW

File: src/app/services/baw-api/base-api.service.spec.ts [hunk 1]
```diff
@@ -65,4 +65,58 @@ describe("BawApiService", () => {
     expect(service.getSessionUser().authToken).toBe("aaaaaaaaaaaaaaaaaaaaaa");
     expect(service.getSessionUser().userName).toBe("username");
   });
+
```
Comment (atruskie) [inline src/app/services/baw-api/base-api.service.spec.ts]: these tests can be moved to a `stringTemplateTests.ts` class

File: src/app/services/baw-api/base-api.service.ts [hunk 2]
```diff
@@ -62,126 +64,72 @@ export abstract class BawApiService {
...
+          .subscribe(this.handleResponse(next, error));
+      case "UPDATE":
+        return this.http
+          .patch<APIResponse>(path, body)
+          .subscribe(this.handleResponse(next, error));
+      case "DELETE":
+        return this.http
+          .delete<APIResponse>(path)
+          .subscribe(this.handleResponse(next, error, true));
+    }
```
Comment (atruskie) [inline src/app/services/baw-api/base-api.service.ts]: revert this change, please. 

Why not use methods?
Stringly type invocation is not fantastic  - especially since you didn't handle any other case in the switch. (I do soee the string-type restriction in the parameter type, but types get erased at compile time).

File: src/app/services/baw-api/base-api.service.ts [hunk 3]
```diff
@@ -62,126 +64,72 @@ export abstract class BawApiService {
...
+   * parameter type compatibility with the placeholders from the interpolated string.
+   * @param strings The strings around the tokens to template
+   * @param placeholders Placeholders that are substituted for token when
+   *  templating is done. Note these are transform functions.
    */
-  protected apiDelete(
-    next: (data: any) => void,
-    error: (err: any) => void,
-    path: string
+  protected makeTemplate<T extends ((any: any) => any)[]>(
```
Comment (atruskie) [inline src/app/services/baw-api/base-api.service.ts]: makeTemplate is general enough to use anywhere. Put it in its own file, perhaps `stringTemplating.ts`

File: src/app/services/baw-api/security.service.ts [hunk 4]
```diff
@@ -23,35 +23,34 @@ export class SecurityService extends BawApiService {
 
     this.loggedInTrigger.next(this.isLoggedIn());
     this.paths = {
-      register: "/security",
-      signIn: "/security",
-      signOut: "/security"
+      register: this.makeTemplate`/security`,
```
Comment (atruskie) [inline src/app/services/baw-api/security.service.ts]: you probably don't _need_ to use make template around a string with no parameters

**Review by atruskie — APPROVED** [review]
Some minor changes required. Factor out `makeTemplate` mainly.

I think you should tackle #78 next.

File: src/app/services/baw-api/security.service.ts (hunk 4 above)
Comment (Allcharles) [reply src/app/services/baw-api/security.service.ts]: This is mostly just for standardising the services before i tackle #78 .

File: src/app/services/baw-api/base-api.service.ts (hunk 2 above)
Comment (Allcharles) [reply src/app/services/baw-api/base-api.service.ts]: My bad, misread https://docs.angularjs.org/api/ngResource/service/$resource and thought it was building a single function which handles all the requests instead of a class.
